crates/communication/src/server/parameters/storage.rs

```diff
@@ -0,0 +1,304 @@
+use std::{path::Path, sync::Arc};
+
+use framework::Writer;
+use serde::{de::DeserializeOwned, Serialize};
+use serialize_hierarchy::{SerializeHierarchy, SerializedValue};
+use tokio::{
+    spawn,
+    sync::{mpsc::Receiver, Notify},
+    task::JoinHandle,
+};
+
+use crate::messages::{ParametersResponse, Response, TextualResponse};
+
+use super::{
+    directory::{deserialize, serialize},
+    StorageRequest,
+};
+
+pub(crate) fn storage<Parameters>(
+    parameters_writer: Writer<Parameters>,
+    parameters_changed: Arc<Notify>,
+    mut request_receiver: Receiver<StorageRequest>,
+    parameters_directory: impl AsRef<Path> + Send + Sync + 'static,
+    body_id: String,
+    head_id: String,
+) -> JoinHandle<()>
+where
+    Parameters: Clone + DeserializeOwned + Send + Serialize + SerializeHierarchy + Sync + 'static,
+{
+    spawn(async move {
+        let mut parameters = (*parameters_writer.next()).clone();
+        while let Some(request) = request_receiver.recv().await {
+            handle_request(
+                request,
+                &mut parameters,
+                &parameters_writer,
+                &parameters_changed,
+                &parameters_directory,
+                &body_id,
+                &head_id,
+            )
+            .await;
+        }
+    })
+}
+
+async fn handle_request<Parameters>(
+    request: StorageRequest,
+    parameters: &mut Parameters,
+    parameters_writer: &Writer<Parameters>,
+    parameters_changed: &Arc<Notify>,
+    parameters_directory: impl AsRef<Path>,
+    body_id: &str,
+    head_id: &str,
+) where
+    Parameters: Clone + DeserializeOwned + Serialize + SerializeHierarchy,
+{
+    match request {
```

Comment:
Here too, every branch in this match statement sends a message and does its own error handling.
I'd propose to have the match statement return a message and send it after, dealing with the send result.

Reply:
Hopefully better now. Let's not overengineer it.